### ISSUE TITLE
question: how to handle request response status

### ISSUE THREAD

**Issue by xenogenesi** [thysultan/dio.js#5]
About `.request`, maybe I'm missing something, `then` and `catch` are working correctly but I never see `err` populated, always empty, and while I get the response object from `then` I can't check the server status which may return an application error (eg: 422 data validation failed).

Currently I'm using a workaround saving `xhr` from `config: function(xhr)` but I guess this is not how is supposed to work, what I'm missing?

**Comment by thysultan** [comment]
I can see the convenience in having it trigger the then/catch block... thinking of adding this as either, 1. 400+ codes pass the xhr object directly to the then handler or 2. pass it to the catch handler.

**Comment by thysultan** [comment]
1. will open a way to do the following

```
dio.request('404').then(function (data) {
	if (data && data.status >= 400) {
		throw data.status;
	} else {

	}
}).catch(function (err) {
	console.log(err);
});
```

and 2.

```
dio.request('404').then(function (data) {
	// silence
}).catch(function (err) {
	console.log(err.status);
});
```

**Comment by xenogenesi** [comment]
I like both but solution 2. is more compact and more *semantically correct*,  errors are handled by `catch` (something users would expect without reading any docutmentation I guess).

I'm not sure `.status` is enough or always populated on network errors, for instance if there isn't a service listening (`net::ERR_CONNECTION_REFUSED`) `xhr.status` is 0.

**Comment by thysultan** [comment]
Yes if the server hangs and doesn't send a response status might be 0 so maybe  `xhr.status !== 0 || xhr.status >= 400` would be better.

**Comment by lukeed** [comment]
Option 2 is the correct way

**Comment by thysultan** [comment]
added this, works like option 2 - [changelog](https://github.com/thysultan/dio.js/blob/master/CHANGELOG.md) for more details.

> Notice that currently status code like 421 422 (used in some API for application level error) are sent through then which wouldn't have access to .status in 2. implementation.

anything >= 400 should find itself in the catch handler, the catch handler can then determine what to do with the response and status in the xhr object, or maybe i'm not getting your point?

**Comment by xenogenesi** [comment]
>= 400 should do, gonna give a try today, thanks!

ps: do you plan to support typescript with a .d.ts?

**Comment by thysultan** [comment]
> ps: do you plan to support typescript with a .d.ts?

@xenogenesi now that you mention it, yes. [backing issue tracker](https://github.com/thysultan/dio.js/issues/6)

**Comment by xenogenesi** [comment]
`.request` seems working as expected, nice! the npm minified version though seems still 5.0.3

> now that you mention it, yes

Good to know, looking forward to it

The issue can be closed for me, thanks again.

**Comment by thysultan** [comment]
@xenogenesi Thanks, published 5.0.5 that fixes that.
